source/common/config/delta_subscription_impl.h

```diff
@@ -0,0 +1,221 @@
+      return;
+    }
+    if (paused_) {
+      ENVOY_LOG(trace, "API {} paused during sendDiscoveryRequest().", type_url_);
+      return;
+    }
+
+    request_.clear_resource_names_subscribe();
+    request_.clear_resource_names_unsubscribe();
+    std::copy(diff.added_.begin(), diff.added_.end(),
+              Protobuf::RepeatedFieldBackInserter(request_.mutable_resource_names_subscribe()));
+    std::copy(diff.removed_.begin(), diff.removed_.end(),
+              Protobuf::RepeatedFieldBackInserter(request_.mutable_resource_names_unsubscribe()));
+
+    ENVOY_LOG(trace, "Sending DiscoveryRequest for {}: {}", type_url_, request_.DebugString());
+    sendMessage(request_);
+    request_.clear_error_detail();
+    request_.clear_initial_resource_versions();
+  }
+
+  void subscribe(const std::vector<std::string>& resources) {
+    ENVOY_LOG(debug, "delta subscribe for " + type_url_);
+
+    // Lazily kick off the requests based on first subscription. This has the
+    // convenient side-effect that we order messages on the channel based on
+    // Envoy's internal dependency ordering.
+    if (!subscribed_) {
+
+      subscribed_ = true;
+    }
+    buildAndQueueDiscoveryRequest(resources);
+  }
+
+  void pause() {
+    ENVOY_LOG(debug, "Pausing discovery requests for {}", type_url_);
+    ASSERT(!paused_);
+    paused_ = true;
+  }
+
+  void resume() {
+    ENVOY_LOG(debug, "Resuming discovery requests for {}", type_url_);
+    ASSERT(paused_);
+    paused_ = false;
+  }
+
+  // Config::SubscriptionCallbacks
+  void onConfigUpdate(const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources,
+                      const Protobuf::RepeatedPtrField<std::string>& removed_resources,
+                      const std::string& version_info) {
+    callbacks_->onConfigUpdate(added_resources, removed_resources, version_info);
+    for (const auto& resource : added_resources) {
+      resources_[resource.name()] = resource.version();
+    }
+    stats_.update_success_.inc();
+    stats_.update_attempt_.inc();
+    stats_.version_.set(HashUtil::xxHash64(version_info));
+    ENVOY_LOG(debug, "Delta config for {} accepted with {} resources added, {} removed", type_url_,
+              added_resources.size(), removed_resources.size());
+  }
+
+  void handleResponse(std::unique_ptr<envoy::api::v2::DeltaDiscoveryResponse>&& message) override {
+    ENVOY_LOG(debug, "Received gRPC message for {} at version {}", type_url_,
+              message->system_version_info());
+
+    request_.set_response_nonce(message->nonce());
+
+    try {
+      onConfigUpdate(message->resources(), message->removed_resources(),
+                     message->system_version_info());
+    } catch (const EnvoyException& e) {
+      stats_.update_rejected_.inc();
+      ENVOY_LOG(warn, "delta config for {} rejected: {}", type_url_, e.what());
+      stats_.update_attempt_.inc();
+      if (callbacks_) {
```

Comment:
In what situation would we not have `callbacks_`? We don't check this on the happy path.

Reply:
Done.